### DISCUSSION TITLE
OpenApi based response checks

### DISCUSSION THREAD

**Discussion by LennSar** [Cornutum/tcases#219] [Ideas]
Hi there,

I wanted to touch on this quote from the documentation:

> That leaves part https://github.com/Cornutum/tcases/pull/4: defining the expected responses. But that's the part of the test program you have to fill in yourself. Tcases for OpenAPI has no way to predict what your API will do for any request, much less a request using random inputs.

I'm pretty sure you know best yourself but I have to disagree on this. We took a lot of effort to write an OpenApi.yaml which covers a lot of information on how the reponse should look like. Of course you cannot tell the exact answer but you could check whether the returned values are in the specified range, according to the defined regex or part of the defined enum. All this could be added during the automatic test case generation based on the OpenApi documentation. SmartBear ReadyApi for example does exactly that. 

So my question would be: Is tcases already capable of doing something like that and I just didn't figure out how to do that yet and if that's not the case: Is this something you think about adding in the future? 

Oh and of course I need to say: Thank you so much to take the time and effort to write this tool and make it public.  As a Java novice I had a real hard time to get it running (and I would have failed if it weren't for the help of a java experiences colleague) but now where it does I realised that this a crazy powerful tool that at the end is much better than many programs you have to pay for. 

Thanks!

Lennart

**Comment by kerrykimbrough** [comment]
Hi, Lennart

You make some very good points. Response validation is certainly a weak spot in Tcases for OpenAPI.

To answer your questions...  Can the current version of TFOA generate code that checks response schema conformance? No, it can't. Could this be added to a future version of TFOA? Yes, I believe that is technically possible, with certain limitations.

One issue is that, in general, response data can be encoded according to many different media types. `application/json` is a common type, but it is not the only possibility. It's not practical for TFOA to support every possible media type. For some media types, it might not even be possible at all.  On the other hand, support for `application/json` might be sufficient for many users. Which media types does your API use?

For response data that uses the `application/json` media type, it's possible for TFOA to generate the kind of response validation code that you suggest. I will definitely consider adding this feature to the next version of TFOA.

**Reply by LennSar** [reply]
Hi Kerry,

thanks for the quick reply. The issue of response data encoding sure does dramatically increase the amount of effort required. Even if it's only done for application/json (which fortunately we do use as well) it wouldn't be a small update to get all the informations required from the OpenApi Documentation. In our case we are using a lot of references which would further increase the required effort.
I'm really happy to read you're consider adding it in the next version of TFOA. If I can be of any help like testing things on our API, please let me know. I'll be happy with what I have so far and stay posted on any updates. 

Thanks,

Lennart

**Comment by kerrykimbrough** [comment]
@LennSar Response validation is here! See [release 3.8.0](https://github.com/Cornutum/tcases/blob/master/ReleaseNotes.md#380).

**Comment by LennSar** [comment]
Hi Kerry,
thanks for the exciting news. You are fast! Really excited to take a look on the update!

Thanks,

Lennart
